# Exploring eBay Car Sales Data
This is a guided project provided by Datarequest, which aims to clean and analyze the dataset of used cars from eBay Kleinanzeigen. The dataset contains 50000 data point, provided with the following columns: 
- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [4]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [5]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

As can be seen from the info method above:
- the columns vehicleType, model,  gearbox, fuelType, and notRepairedDamage are not of total 50000. 
- some of the data are in object which means it is a string and some are in integer values. 
- some numeric data are assigned as object (string) and it can be numeric, like price and odometer.
- lastly, the columns name are found to be in camelcase rather than snakecase. 

## Data Cleaning

### 1) change from camelcase to snakecase

First will change the columns names so it suits the Python's preferred snakecase. 

In [6]:
# columns before
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.columns = ["date_crawled", "name", "seller","offer_type", 
                 "price","ab_test", "vehicle_type", "registration_year",
                 "gearbox", "power_PS", "model", "odometer", "reqistration_month",
                 "fuel_type", "brand","unrepaired_damage", "ad_created", 
                 "pictures_num", "postal_code", "last_seen"
            ]

In [8]:
# columns after
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  reqistration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  pictures_num  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### 2) Translate data values from German to English

Some of the columns are found to be in German words. To be more readable I will change it to English. The columns are: vehicleType, gearbox, fuelType, notRepairedDamage, offerType

In [9]:
# vehicle type

autos["vehicle_type"].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [10]:
vt_dic = {'bus':'bus', 'limousine':'limousine', 'kleinwagen':'supermini', 
          'kombi':'transporter','coupe':'coupe', 'suv':'suv','cabrio':'convertable',
         'andere':'other'}

autos["vehicle_type"] = autos["vehicle_type"].map(vt_dic)
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'supermini', 'transporter', nan, 'coupe',
       'suv', 'convertable', 'other'], dtype=object)

In [11]:
# gearbox

autos["gearbox"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [12]:
gb_dic = {'manuell':'manual', 'automatik':'automatic'}

autos["gearbox"] = autos["gearbox"].map(gb_dic)
autos["gearbox"].unique()

array(['manual', 'automatic', nan], dtype=object)

In [13]:
# offer type

autos["offer_type"].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [14]:
ot_dic = {'Angebot':'Offer', 'Gesuch':'Petition'}

autos["offer_type"] = autos["offer_type"].map(ot_dic)
autos["offer_type"].unique()

array(['Offer', 'Petition'], dtype=object)

In [15]:
# fuel type 

autos["fuel_type"].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [16]:
ft_dic = {'lpg':'lpg', 'benzin':'petrol', 'diesel':'diesel', 
          'cng':'cng','hybrid':'hybrid', 'elektro':'electro','andere':'other'}

autos["fuel_type"] = autos["fuel_type"].map(ft_dic)
autos["fuel_type"].unique()

array(['lpg', 'petrol', 'diesel', nan, 'cng', 'hybrid', 'electro',
       'other'], dtype=object)

In [17]:
# unrepaired_damage 

autos["unrepaired_damage"].unique()

array(['nein', nan, 'ja'], dtype=object)

In [18]:
ud_dic = {'nein':'no', 'ja':'yes'}

autos["unrepaired_damage"] = autos["unrepaired_damage"].map(ud_dic)
autos["unrepaired_damage"].unique()

array(['no', nan, 'yes'], dtype=object)

## Data exploration



In [19]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat      Offer     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gearbox      power_PS  model  \
count         44905       50000.000000   47320  50000.000000  47242   
unique            8                NaN       2           NaN    245   
top       limousine                NaN  manual           NaN   golf   
freq          12859                NaN   36993           NaN   4024   
mean            NaN        2005.073280     NaN    116.355920    NaN   
std             NaN         105.712813     NaN    209.216627    NaN   
min             NaN        1000.000000     NaN      0.000000    NaN   
25%             NaN        1999.000000     NaN     70.000000    NaN   
50%             NaN        2003.000000     NaN    105.000000    NaN   
75%             NaN        2008.000000     NaN    150.000000    NaN   
max             NaN        9999.000000     NaN  17700.000000    NaN   

         odometer  reqistration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    petrol  volkswagen                no   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pictures_num   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

As found earlier, not all numeric data can be shown here. Therefore, these columns will need to be cleaned and stored as numeric

In [20]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",","").astype(int)
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)


### 1) visualizing the price and mileage columns



In [22]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [19]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts())


(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850   

As seen above, there are definitly lots of outliers in the price column, therefore I will limit the price between 500 and 400000

In [20]:
autos = autos[autos["price"].between(500,400000)]

### 2) Calculate the distribution of date values

The columns date_crawled, last_seen, and ad_created were found to be as Strings, and in full timestamp representation

In [21]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In order to understand the date range, I will extract the date values from every column. 

In [22]:
# date_crawled
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025567
2016-03-06    0.014125
2016-03-07    0.036189
2016-03-08    0.033173
2016-03-09    0.032907
2016-03-10    0.032707
2016-03-11    0.033018
2016-03-12    0.037320
2016-03-13    0.015522
2016-03-14    0.036300
2016-03-15    0.034016
2016-03-16    0.029359
2016-03-17    0.031155
2016-03-18    0.012883
2016-03-19    0.034747
2016-03-20    0.038073
2016-03-21    0.037741
2016-03-22    0.033018
2016-03-23    0.032397
2016-03-24    0.028982
2016-03-25    0.031089
2016-03-26    0.032641
2016-03-27    0.031177
2016-03-28    0.034836
2016-03-29    0.033262
2016-03-30    0.033328
2016-03-31    0.031665
2016-04-01    0.033905
2016-04-02    0.035767
2016-04-03    0.038827
2016-04-04    0.036610
2016-04-05    0.013172
2016-04-06    0.003171
2016-04-07    0.001353
Name: date_crawled, dtype: float64

In [23]:
# ad_created
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
2015-12-30    0.000022
2016-01-03    0.000022
2016-01-07    0.000022
2016-01-10    0.000044
2016-01-13    0.000022
2016-01-14    0.000022
2016-01-16    0.000022
2016-01-22    0.000022
2016-01-27    0.000067
2016-01-29    0.000022
2016-02-01    0.000022
2016-02-02    0.000044
2016-02-05    0.000044
2016-02-07    0.000022
2016-02-08    0.000022
2016-02-09    0.000022
2016-02-11    0.000022
2016-02-12    0.000044
2016-02-14    0.000044
2016-02-16    0.000022
2016-02-17    0.000022
2016-02-18    0.000044
2016-02-19    0.000067
2016-02-20    0.000044
2016-02-21    0.000044
                ...   
2016-03-09    0.032996
2016-03-10    0.032441
2016-03-11    0.033328
2016-03-12    0.037098
2016-03-13    0.016963
2016-03-14    0.034880
2016-03-15    0.033794
2016-03-16    0.029847
2016-03-17    0.030822
2016-03-18    0.013504
2016-03-19    0.033616
2016-03-20    0.038207
2016-03-21 

In [24]:
# last_seen
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001087
2016-03-06    0.004169
2016-03-07    0.005211
2016-03-08    0.007007
2016-03-09    0.009468
2016-03-10    0.010289
2016-03-11    0.012041
2016-03-12    0.023904
2016-03-13    0.008870
2016-03-14    0.012285
2016-03-15    0.015677
2016-03-16    0.016165
2016-03-17    0.027674
2016-03-18    0.007406
2016-03-19    0.015411
2016-03-20    0.020423
2016-03-21    0.020667
2016-03-22    0.021243
2016-03-23    0.018405
2016-03-24    0.019536
2016-03-25    0.018582
2016-03-26    0.016476
2016-03-27    0.015456
2016-03-28    0.020534
2016-03-29    0.021354
2016-03-30    0.024148
2016-03-31    0.023438
2016-04-01    0.022862
2016-04-02    0.024880
2016-04-03    0.024946
2016-04-04    0.024303
2016-04-05    0.126616
2016-04-06    0.225314
2016-04-07    0.134155
Name: last_seen, dtype: float64

As can be seen above, the date_crawled and last_seen dates are nearly the same since it's both added by the crawler.

In [25]:
autos["registration_year"].describe()

count    45097.000000
mean      2005.064173
std         89.652017
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

here in the registration_year column, there exist an outliers, the minimum value is 1000 and the maximum is 9999. This should be fixed to appropriate time, let's say between 1900 - 2016

In [26]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    43323.000000
mean      2003.227270
std          7.075229
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

so the minimum and maximum now make more sense. 1774 data were removed.

### 3) Aggregation of brand column

lets aggregate the top 10 brands and find the mean price and mean mileage

In [27]:
# found brands
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'audi', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'mitsubishi', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [28]:
autos["brand"].value_counts()[0:10]

volkswagen       9180
bmw              4983
mercedes_benz    4430
opel             4347
audi             3921
ford             2823
renault          1902
peugeot          1272
fiat             1019
seat              775
Name: brand, dtype: int64

It seems that the cars with brand volkswagen are the most found in the dataset. Now I will calculate the mean price for all the brands

In [29]:
brands_mean_price = {}
brands_mean_mileage = {}
brands = autos["brand"].value_counts()[0:10]

for i in range(0,10): # b for brand and c for count
    selected_rows = autos[autos["brand"] == brands.index[i]]
    mean_price = selected_rows["price"].mean()
    mean_mileage = selected_rows["odometer_km"].mean()
    brands_mean_price[brands.index[i]] = int(mean_price)
    brands_mean_mileage[brands.index[i]] = int(mean_mileage)
    
bmp_series = pd.Series(brands_mean_price)
bmm_series = pd.Series(brands_mean_mileage)

In [30]:
df = pd.DataFrame(bmp_series, columns=["mean_price"])
df["mean_mileage"] = bmm_series
df

mean_price  mean_mileage
audi                 9613        128909
bmw                  8582        132865
fiat                 3256        114416
ford                 4291        123494
mercedes_benz        8766        131019
opel                 3394        128012
peugeot              3360        126073
renault              2819        126351
seat                 4810        120058
volkswagen           5783        128234

and we can say from the top 10 brands, audi, mercedez benz, and bmw was found to be the highest in price, while renault, fiat, and peugeot are the lowest

### 4) Find most common brand/model combinations

In [36]:
brand_model = autos['brand']+'_'+autos['model']
brand_model.value_counts().head(10)

volkswagen_golf           3437
bmw_3er                   2519
volkswagen_polo           1347
volkswagen_passat         1307
opel_corsa                1266
opel_astra                1224
audi_a4                   1199
mercedes_benz_c_klasse    1122
bmw_5er                   1115
mercedes_benz_e_klasse     945
dtype: int64